torch/csrc/jit/passes/freeze_module.cpp

```diff
@@ -0,0 +1,395 @@
+            if (auto attrVal = tryInsertConstant(*graph, attr)) {
+              paramConst = *attrVal;
+            } else {
+              GRAPH_DEBUG(
+                  attr.type()->cast<ClassType>() ? "" : "attribute: ",
+                  name,
+                  " is not materializable.");
+              continue;
+            }
+            std::string fullName("self.");
+            for (auto& name : names_) {
+              fullName += name + '.';
+            }
+            fullName += name;
+            paramConst->setDebugName(fullName);
+            attrValues[attrModule._ivalue()][name] = paramConst;
+          }
+          GRAPH_UPDATE(
+              "Folding GetAttr %",
+              n->outputs()[0]->debugName(),
+              " with ",
+              paramConst->debugName());
+          n->outputs().at(0)->replaceAllUsesWith(paramConst);
+          n->removeAllInputs();
+        }
+      }
+    }
+  }
+
+  // cleanupFrozenModule function cleans up the Frozen module. it performs the
+  // following:
+  // 1) Remove unused attributes.
+  // 2) Remove unreferenced submodules
+  // 3) Remove non public unreferenced methods.
+  // TODO: do #3 because there is no API to 'unsafely' remove methods.
+  void cleanupFrozenModule(std::shared_ptr<Graph>& graph) {
+    std::vector<std::string> attrsToRemove;
+    auto type = module_.type();
+    size_t N = type->numAttributes();
+    recordReferencedAttrs(graph);
+    auto it = preservedAttrs_.find(module_._ivalue());
+    auto it2 = preservedScalarAttrs_.find(module_._ivalue());
+    for (size_t i = 0; i < N; ++i) {
+      auto attrTy = type->getAttribute(i);
+      auto name = type->getAttributeName(i);
+      auto attr = module_.attr(name);
+      bool immutable;
+      if (AliasDb::mutableType(attrTy)) {
+        immutable = it == preservedAttrs_.end() || !it->second.count(attr);
+      } else {
+        immutable =
+            it2 == preservedScalarAttrs_.end() || !it2->second.count(name);
+      }
+      if (immutable) {
+        attrsToRemove.push_back(name);
+      }
+    }
+    for (auto& name : attrsToRemove) {
+      module_._ivalue()->unsafeRemoveAttr(name);
+      module_.type()->unsafeRemoveAttribute(name);
+    }
+    for (auto& fn : type->methods()) {
+      auto& name = fn->name();
+      if ("forward" == name)
+        continue;
+      type->unsafeRemoveMethod(name);
+      module_._ivalue()->compilation_unit()->unsafeRemoveMethod(fn->qualname());
+    }
+  }
+
+  // Contains attributes that can't be folded or user directs to keep them.
+  std::unordered_map<script::ModulePtr, IValue::HashAliasedIValues>
+      preservedAttrs_;
+  // Tracked immutable types (Scalars) by their attribute names not
+  // IValues.
+  std::unordered_map<script::ModulePtr, std::unordered_set<std::string>>
+      preservedScalarAttrs_;
+
+  script::Module& module_;
+
+  // Contains the attributes names (e.g. {"self", "subModule", "a"}
+  std::deque<std::string> names_;
```

Comment:
Not sure why this is a class attribute, probably shouldn't be

Reply:
I used it to pretty print (e.g.  %self.sub.conv = prim::constant(...). the time of computing the name and inserting the new node are far apart.